tempesta_fw/tls.c

```diff
@@ -666,6 +703,84 @@ tfw_tls_cfg_require(void)
 	tfw_tls_cgf |= TFW_TLS_CFG_F_REQUIRED;
 }
 
+int
+tfw_tls_cfg_alpn_protos(const char *cfg_str)
+{
+	ttls_alpn_proto *protos;
+	size_t len = strlen(cfg_str);
+	int order = 0;
+
+#define CONF_HTTPS	"https"
+#define CONF_HTTP1	"http1"
+#define CONF_HTTP2	"http2"
+#define CONF_LEN	5
+	if (strncasecmp(cfg_str, CONF_HTTPS, CONF_LEN))
+		return -EINVAL;
+
+	if (len == CONF_LEN)
+		goto found;
+
+	cfg_str += CONF_LEN;
+	if (cfg_str[0] != ':')
+		return -EINVAL;
+
+	++cfg_str;
+	len -= CONF_LEN + 1;
+	if (len >= CONF_LEN && !strncasecmp(cfg_str, CONF_HTTP1, CONF_LEN))
+	{
+		cfg_str += CONF_LEN;
+		len -= CONF_LEN;
+		if (!len || (cfg_str[0] == ','
+			     && !strcasecmp(++cfg_str, CONF_HTTP2)))
+		{
+			goto found;
+		}
+	}
+	else if (len >= CONF_LEN &&
+		 !strncasecmp(cfg_str, CONF_HTTP2, CONF_LEN))
+	{
+		order = 1;
+		cfg_str += CONF_LEN;
+		len -= CONF_LEN;
+		if (!len || (cfg_str[0] == ','
+			     && !strcasecmp(++cfg_str, CONF_HTTP1)))
+		{
+			goto found;
+		}
+	}
+
+	return -EINVAL;
+
+found:
+	protos = kzalloc(TTLS_ALPN_PROTOS * sizeof(ttls_alpn_proto), GFP_ATOMIC);
+	if (unlikely(!protos))
+		return -ENOMEM;
+
+	protos[order].name = TTLS_ALPN_HTTP1;
+	protos[order].len = sizeof(TTLS_ALPN_HTTP1) - 1;
+	protos[order].id = TTLS_ALPN_ID_HTTP1;
+	protos[!order].name = TTLS_ALPN_HTTP2;
+	protos[!order].len = sizeof(TTLS_ALPN_HTTP2) - 1;
+	protos[!order].id = TTLS_ALPN_ID_HTTP2;
```

Comment:
Looks like HTTP/2 is unconditionally enabled even if `listen` directive has `https:http1` and doesn't mention http2 at all. Is this intended behavior?

Reply:
Yes, currently `http1` is just a shortcut for `http1,http2` and `http2` is a shortcut for `http2,http1`. Thus, we always enable http2 handling, and since our protocols order takes precedence, we can define what protocol will be used when client sends us in any order both `http/1.1` and `h2` in ALPN.
But in any case, this point (as well as the names for HTTP/1.1 and HTTP/2 protocols in configuration) remains the subject TBD.

Reply:
OK. Corrected.